common/src/test/java/org/conscrypt/javax/net/ssl/SSLSocketVersionCompatibilityTest.java

```diff
@@ -1446,6 +1446,52 @@ public Void call() throws Exception {
         server.close();
     }
 
+    /**
+     * Test to confirm that an SSLSocket.close() on one
+     * thread will interrupt another thread blocked writing on the same
+     * socket.
+     *
+     * See also b/147323301 where close() triggered an infinite loop instead.
+     */
+    @Test
+    public void test_SSLSocket_interrupt_write_withAutoclose() throws Exception {
+        final TestSSLContext c = new TestSSLContext.Builder()
+            .clientProtocol(clientVersion)
+            .serverProtocol(serverVersion)
+            .build();
+        final Socket underlying = new Socket(c.host, c.port);
+        final SSLSocket wrapping = (SSLSocket) c.clientContext.getSocketFactory().createSocket(
+            underlying, c.host.getHostName(), c.port, true);
+        final byte[] data = new byte[1024 * 220];
+
+        Future<Void> clientFuture = runAsync(new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+                wrapping.startHandshake();
+                try {
+                    wrapping.getOutputStream().write(data);
+                    // Failure here means that no exception was thrown, so the data buffer is
+                    // probably too small.
+                    fail();
+                } catch (SocketException expected) {
+                    assertTrue(expected.getMessage().contains("closed"));
```

Comment:
Asserting that message is "Socket closed" probably better since "closed" could match other SocketException that is thrown in waitForHandshake()

Reply:
Yeah, you'd think that :)

But there are a number of places in the code where these can be generated, including from the underlying SocketOutputStream (which is platform dependent), and they generate a mix of "Socket closed", "Socket is Closed", and "Socket is closed" messages.  I'll double whether check the "Socket is Closed" one is possible from this test and if so make it case insensitive...

Also, not planning to merge until I guess the test passing consistently across Android, OpenJDK and Travis CI which will probably mean guessing a buffer size which works consistently everywhere.

Reply:
Ah sorry, I see your point.  Yeah, this test doesn't guarantee that the exception will be thrown by the new code in writeInternal() as it's very timing and platform dependent as I mentioned in the description.

I have however on both Android and OpenJDK manually observed it spinlooping on close without the new conditional in writeInternal() and throwing from writeInternal when the new code is added.

So whilst it's not a perfect regression test, it is a reasonable test for the overall behaviour, which is that both implementations of SSLSocket should throw the same exception when an SSLSocket is closed mid-write by another thread.